common/src/ipc/serde.rs

```diff
@@ -0,0 +1,631 @@
+use std::mem::size_of;
+use std::mem::size_of_val;
+use std::num::NonZeroI32;
+use std::num::NonZeroU8;
+use std::time::Duration;
+
+use super::types2::Animation;
+use super::types2::BitPack;
+use super::types2::ClearRequest;
+use super::types2::Coord;
+use super::types2::Image;
+use super::types2::ImageDescription;
+use super::types2::ImageRequest;
+use super::types2::Info;
+use super::types2::PixelFormat;
+use super::types2::Scale;
+use super::types2::Transition;
+use super::types2::TransitionType;
+use super::types2::Vec2;
+
+/// Type for managing access in [`Serialize`] and [`Deserialize`]
+pub struct Cursor<T> {
```

Comment:
This seems nearly identical to `std::io::Cursor` (just with a usize vs a u64) - why don't we just use that?

Reply:
I looked at it, but didn't see a reason to use it. It implements `BufRead` and `Write`, but I would want to panic anyway with the way I allocate memory (with `size` on `Serialize`). I would also implement extension traits to ease tagged slice writing, so I just wrote couple of lines and left it at that.

`io::Cursor` was the first thing I looked at, but disliked being unable to directly access it.

Reply:
Sorry, I don't know if I'm following - `io::Cursor` has `get_ref()` and `get_ref_mut()` methods to get a reference to the underlying data, so you can directly access it.

And are you saying that when writing a wrong amount of data to the `Cursor` you'd rather panic than get back information about how much was written? That would make sense as a rational, I guess, but feels like less ideal for a user in the long run.

Reply:
If I read correctly, `get_*` gets underlying value, not remaining. There is `remaining_slice`, which is nightly for some reason. So I just implemented bare minimum of what would be required.

And with the panics, given that (currently) reading and writing is constrained to `common` crate, I wanted to implement fuzzing for correctness. So I just opted for panics.

Reply:
> And are you saying that when writing a wrong amount of data to the Cursor you'd rather panic than get back information about how much was written? That would make sense as a rational, I guess, but feels like less ideal for a user in the long run.

I disagree with the last part; I think it makes *more* sense for the user to panic. This isn't a recoverable error anyway: if our serialization/deserialization fails, something fundamentally wrong has happened (for example, maybe a different program is trying to write something malicious to the `swww` socket). It also makes for a much easier-to-debug error, since it's just a panic.